QuEST/src/CPU/QuEST_cpu.c

```diff
@@ -942,6 +942,40 @@ qreal densmatr_calcHilbertSchmidtDistanceSquaredLocal(Qureg a, Qureg b) {
     return trace;
 }
 
+/** computes Tr(conjTrans(a) b) = sum of (a^* b) */
+qreal densmatr_calcHilbertSchmidtScalarProductLocal(Qureg a, Qureg b) {
+    
+    long long int index;
+    long long int numAmps = a.numAmpsPerChunk;
+        
+    qreal *aRe = a.stateVec.real;
+    qreal *aIm = a.stateVec.imag;
+    qreal *bRe = b.stateVec.real;
+    qreal *bIm = b.stateVec.imag;
+    
+    qreal trace = 0, traceIm = 0;
+    
+# ifdef _OPENMP
+# pragma omp parallel \
+    shared    (aRe,aIm, bRe,bIm, numAmps) \
+    private   (index, traceIm) \
+    reduction ( +:trace )
+# endif 
+    {
+# ifdef _OPENMP
+# pragma omp for schedule  (static)
+# endif
+        for (index=0LL; index<numAmps; index++) {
+            trace += aRe[index]*bRe[index] + aIm[index]*bIm[index];
+            traceIm += aRe[index]*bIm[index] - aIm[index]*bRe[index];
+        }
+    }
+    
+    if(abs(traceIm) > REAL_EPS) printf("ERROR: too large imag. part of calcHilbertSchmidtScalarProduct");
```

Comment:
I think `traceIm` shouldn't even be computed; a pre-condition is that density matrices `a` and `b` are Hermitian to numerical precision, and ergo `traceIm` should be zero. Users are warned in the doc for using Hermitivity-breaking functions, and this (outside of `validation`) isn't the right place to "throw" an error.

Reply:
Sure, this part of the code was irrelevant -- I removed it in bb79977ff0cddaf55527bb0dc1221e993bd466ef